lightning/src/ln/channelmanager.rs

```diff
@@ -1206,108 +1223,154 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref> ChannelMan
+	/// May generate SendHTLCs message(s) event on success, which should be relayed.
 	///
-	/// Raises APIError::RoutError when invalid route or forward parameter
-	/// (cltv_delta, fee, node public key) is specified.
-	/// Raises APIError::ChannelUnavailable if the next-hop channel is not available for updates
-	/// (including due to previous monitor update failure or new permanent monitor update failure).
-	/// Raised APIError::MonitorUpdateFailed if a new monitor update failure prevented sending the
-	/// relevant updates.
+	/// Each path may have a different return value, and PaymentSendValue may return a Vec with
+	/// each entry matching the corresponding-index entry in the route paths.
 	///
-	/// In case of APIError::RouteError/APIError::ChannelUnavailable, the payment send has failed
-	/// and you may wish to retry via a different route immediately.
-	/// In case of APIError::MonitorUpdateFailed, the commitment update has been irrevocably
-	/// committed on our end and we're just waiting for a monitor update to send it. Do NOT retry
-	/// the payment via a different route unless you intend to pay twice!
+	/// In general, a path may raise:
+	///  * APIError::RouteError when an invalid route or forwarding parameter (cltv_delta, fee,
+	///    node public key) is specified.
+	///  * APIError::ChannelUnavailable if the next-hop channel is not available for updates
+	///    (including due to previous monitor update failure or new permanent monitor update
+	///    failure).
+	///  * APIError::MonitorUpdateFailed if a new monitor update failure prevented sending the
+	///    relevant updates.
+	///
+	/// Note that depending on the type of the PaymentSendFailure the HTLC may have been
+	/// irrevocably committed to on our end. In such a case, do NOT retry the payment with a
+	/// different route unless you intend to pay twice!
 	///
 	/// payment_secret is unrelated to payment_hash (or PaymentPreimage) and exists to authenticate
 	/// the sender to the recipient and prevent payment-probing (deanonymization) attacks. For
 	/// newer nodes, it will be provided to you in the invoice. If you do not have one, the Route
 	/// must not contain multiple paths as otherwise the multipath data cannot be sent.
 	/// If a payment_secret *is* provided, we assume that the invoice had the basic_mpp feature bit
 	/// set (either as required or as available).
-	pub fn send_payment(&self, route: Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>) -> Result<(), APIError> {
-		if route.paths.len() < 1 || route.paths.len() > 1 {
-			return Err(APIError::RouteError{err: "We currently don't support MPP, and we need at least one path"});
+	pub fn send_payment(&self, route: Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>) -> Result<(), PaymentSendFailure> {
+		if route.paths.len() < 1 {
+			return Err(PaymentSendFailure::ParameterError(APIError::RouteError{err: "There must be at least one path to send over"}));
 		}
```

Comment:
Prior to this commit, why was `APIError::RouteError` used instead of `APIError::APIMisuseError` for these sort of errors?

I'm wondering if this function's interface can remain consistent with the rest of the API (i.e., still return `Result<(), APIError>`) by using `APIError::APIMisuseError` for these errors and then repurposing `APIError::RouteError` to cover the remaining variants of `PaymentSendFailure`.

Reply:
> A user may have a general APIError handling function, which could be pretty common, but send_payment requires a pretty different handling.

AFAICT, `RouteError` is only used by `send_payment` so any handling would already be specific to it, right?

> Moreover, several of the APIError values (especially ChannelUnavailable and MonitorUpdateFailed) need to be returned here per-path, not in general, so it feels awkward to include them in one of the APIError enums.

Are you saying that because `send_payment` could return simply `APIError::ChannelUnavailable` (even though it should never do so), then the interface shouldn't use `Result<(), APIError>` for its return type?

If so, how's this different from another function using `APIError` for its return type even though it should never return `APIError::RouteError`?

I don't mean to come across as overly pushy about this. Just trying to tease out the intended usage of `APIError` and whether its original purpose is no longer suitable for our needs.

FWIW, I had been thinking something more like this (naming and exact definition TBD):

```rust

pub enum PaymentFailure {
    PathParameterError(Vec<Result<(), APIError>>),
    AllFailedRetrySafe(Vec<APIError>),
    PartialFailure(Vec<Result<(), APIError>>)
}

pub enum APIError {
    APIMisuseError { },
    FeeRateTooHigh { },
    RouteError {
        err: PaymentFailure,
    },
    ChannelUnavailable { },
    MonitorUpdateFailed
}

pub fn send_payment(&self, route: Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>) -> Result<(), APIError> { }
```

Then `send_payment` could return an `APIError` other than `RouteError` for path-independent errors and `RouteError` otherwise. With the caveat of possibly better naming for these.

That said, I'm more concerned with my more high-level point on whether `APIError` is still suitable rather than specifically advocating for this approach.

> It my feel a bit more natural to remove the APIError that is in ParameterError, since its always a RouteError, though - would that address your unease?

Possibly. I'm still not quite sure I understand the difference between `APIMisuseError` and `RouteError`, though,  given the latter is only used by `send_payment`. Seems like `RouteError` is essentially "`APIMisuseError` but for send_payment".

Reply:
Right, I think I agree that APIError has gotten pretty awkward - its better than what was there previously (just a general error that was used throughout the codebase), but its still not great. As for your above suggestion, that feels maybe even more awkward - now we have an APIError inside a struct inside an APIError. What about something without RouteError at all, with the appropriate str? We could even drop APIError completely and just expose a part of it in PaymentFailure (with just FeeRateTooHigh, ChannelUnavailable, MonitorUpdateFailed).

```
pub enum PaymentFailure {
    ParameterError(&'static str),
    PathParameterError(Vec<Result<(), &'static str>>),
    AllFailedRetrySafe(Vec<APIError>),
    PartialFailure(Vec<Result<(), APIError>>)
}

pub enum APIError {
    APIMisuseError { },
    FeeRateTooHigh { },
    ChannelUnavailable { },
    MonitorUpdateFailed
}

pub fn send_payment(&self, route: Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>) -> Result<(), PaymentFailure> { }
```

Reply:
I'm ok with dropping `RouteError` as you described and using `PaymentFailure` as the error half of the return type. Note that `build_onion_payloads` in onion_utils.rs also uses `RouteError`.

As discussed offline, we can circle back later on what should become of `APIError`. I'll open an issue for discussion.

Reply:
Opened #576 for further discussion.

Reply:
Oh, ugh, right, my solution doesn't actually work - we need to sometimes return a per-path error due to invalid keys along the path, which needs to be an APIError and is a "Route Error".